### PR TITLE
nvim_buf_*() and unloaded buffers

### PR DESCRIPTION
**Pull request by phodge**
Proposed solution for #7688 and #5534. Changes in this PR:

* `nvim_buf_get_lines()` will now always return an empty list for an unloaded buffer, instead of a list containing a single empty string.
* `nvim_buf_line_count()` will now always return `0` for an unloaded buffer, instead of `1`.
* `nvim_buf_is_loaded()` has been added to check if a buffer is loaded
* help text updates for the above
* extra unit tests for the above

### CODE REVIEW

File: runtime/doc/api.txt [hunk 1]
```diff
@@ -847,7 +847,8 @@ nvim_buf_line_count({buffer})                          *nvim_buf_line_count()*
                     {buffer}  Buffer handle
 
                 Return:~
-                    Line count
+                    Line count, or `0` if the buffer has been unloaded (see
```
Comment (bfredl) [inline runtime/doc/api.txt]: This section is autogenerated, please change the docstrings in api/buf.c only

Comment (phodge) [reply runtime/doc/api.txt]: What command can I run to regenerate the api docs locally?

Comment (jamessan) [reply runtime/doc/api.txt]: `python3 scripts/gen_api_vimdoc.py`

File: src/nvim/api/buffer.c [hunk 2]
```diff
@@ -745,10 +758,34 @@ void nvim_buf_set_name(Buffer buffer, String name, Error *err)
...
+
+/// Checks if a buffer is valid.
+///
+/// @note Even if a buffer is valid it may have been unloaded. Use
+///       |nvim_buf_is_loaded()| to check if a buffer is loaded.
 ///
 /// @param buffer Buffer handle
-/// @return true if the buffer is valid, false otherwise
+/// @return true if the buffer is valid, false otherwise.
 Boolean nvim_buf_is_valid(Buffer buffer)
```
Comment (justinmk) [inline src/nvim/api/buffer.c]: Just noticed that this returns true for `0`.

Comment (phodge) [reply src/nvim/api/buffer.c]: Yeah all of the `nvim_buf_*()` functions accept a buffer id of `0`, which is interpreted as "the current buffer". This feature is used in the lua tests.

**Comment by phodge** [comment]
The QuickBuild error (from inccommand tests) appears to be unrelated?

**Comment by KillTheMule** [comment]
That's the "inccomand during term activity" test, that's known to be flakey.

File: src/nvim/api/buffer.c [hunk 3]
```diff
@@ -745,10 +758,34 @@ void nvim_buf_set_name(Buffer buffer, String name, Error *err)
...
+///
+/// Buffers may have been unloaded by the |:bunload| command or
+/// the buffer's|'bufhidden'| option.
+///
+/// @param buffer Buffer handle
+/// @return true if the buffer is valid and loaded, false otherwise.
+Boolean nvim_buf_is_loaded(Buffer buffer)
+  FUNC_API_SINCE(5)
+{
+  Error stub = ERROR_INIT;
```
Comment (bfredl) [inline src/nvim/api/buffer.c]: Maybe we should use `Error *err` here, then it would be possible to identify the three states: loaded, unloaded (thus loadable) and invalid in one RPC call.

File: src/nvim/api/buffer.c (hunk 2 above)
Comment (justinmk) [reply src/nvim/api/buffer.c]: Yes but does it make sense for this function?

Comment (bfredl) [reply src/nvim/api/buffer.c]: @justinmk what would 0 mean instead in this case? And why is that better than curbuf?

Comment (justinmk) [reply src/nvim/api/buffer.c]: @bfredl I don't understand your question about curbuf. 

I'm thinking `nvim_buf_is_valid(0)` should return false, because `nvim_buf_is_valid` is for checking validity of buffer handles.

Comment (bfredl) [reply src/nvim/api/buffer.c]: But 0 is a valid buffer handle, and it means curbuf. Why should it mean something different for this function?

Comment (justinmk) [reply src/nvim/api/buffer.c]: It's a special value that only is relevant in the context of a buffer-related function. If someone uses nvim_buf_is_valid to validate a loop or something like that, then `nvim_buf_is_valid (0) => true`  might be unhelpful.

Comment (bfredl) [reply src/nvim/api/buffer.c]: And exactly how is this function not even _related_ to buffers? I'm not saying 1 is obviously objectively better _a priori_, just that if we are introducing a special case, and a change of behaviour of existing input at that, I want a stronger argument than "maybe false is more useful sometimes". After all, clients can already special case 0 as they want, the _only_ functional effect of changing this now is potentially breaking code that already uses this function and didn't anticipate this change.

Comment (phodge) [reply src/nvim/api/buffer.c]: If neovim was ever refactored such that the `curbuf` global variable could be NULL, perhaps during early startup when neovim hasn't had a chance to open a file yet, then `nvim_buf_is_valid(0)` would be the logical way to ask neovim "Is there a current buffer yet?".

Comment (bfredl) [reply src/nvim/api/buffer.c]: @phodge In that hypothetical situation: yes. Then most other `nvim_buf_` functions would also unconditionally throw an error in that situation. (if we _want_ such a situation to be exposed is of course a different question, the programming model is simpler as long as we can avoid it)

File: src/nvim/api/buffer.c (hunk 3 above)
Comment (phodge) [reply src/nvim/api/buffer.c]: My hesitation with returning an error for invalid buffers is that you'd now need to introduce error handling code (try/except in python or pcall() in lua) to use buf.is_loaded(). Which in turn means you can't write a one-liner like `if buf.is_loaded() && buf.line_count() > 10:`

Comment (bfredl) [reply src/nvim/api/buffer.c]: That makes sense. We could add some more detailed state function later it need be.

**Comment by phodge** [comment]
@justinmk, @bfredl: I'd like to finalize this PR if possible, is there anything else that needs addressing?

**Comment by phodge** [comment]
I've updated the description to mention that #5534 will also be resolved by this PR.

File: src/nvim/api/buffer.c [hunk 4]
```diff
@@ -35,7 +35,8 @@
 ///
 /// @param buffer   Buffer handle
 /// @param[out] err Error details, if any
-/// @return Line count
+/// @return Line count, or \`0` if the buffer has been unloaded (see
+///         |nvim_buf_is_loaded()|).
```
Comment (bfredl) [inline src/nvim/api/buffer.c]: perhaps this could be formulated to make it clear if `nvim_buf_line_count` is already used, then `nvim_buf_is_loaded` need not also be used. The later is only referenced because shared info of all buffer functions is placed there, ideally this information could be at `|api-buffer|` before any specific function (but the generation script maybe needs to be changed to not overwrite such a section).

File: src/nvim/api/buffer.c (hunk 3 above)
Comment (bfredl) [reply src/nvim/api/buffer.c]: (for posterity: just realized `nvim_buf_line_count` is already this function.)

File: test/functional/api/buffer_spec.lua [hunk 5]
```diff
@@ -35,8 +35,42 @@ describe('api/buf', function()
...
+    it('line_count has defined behaviour for unloaded buffers', function()
+      -- we'll need to know our bufnr for when it gets unloaded
+      local bufnr = request('nvim_buf_get_number', 0)
+      -- replace the buffer contents with these three lines
+      request('nvim_buf_set_lines', bufnr, 0, -1, 1, {"line1", "line2", "line3", "line4"})
+      -- check the line count is correct
+      eq(4, request('buffer_line_count', bufnr))
+      eq(4, request('nvim_buf_line_count', bufnr))
+      -- force unload the buffer (this will discard changes)
+      request('nvim_command','new')
```
Comment (bfredl) [inline test/functional/api/buffer_spec.lua]: use `command('new')`

File: test/functional/api/buffer_spec.lua [hunk 6]
```diff
@@ -35,8 +35,42 @@ describe('api/buf', function()
...
+      request('nvim_command', 'bunload! '..bufnr)
+      -- line count for an unloaded buffer should always be 0
+      eq(0, request('buffer_line_count', bufnr))
```
Comment (bfredl) [inline test/functional/api/buffer_spec.lua]: obsolete functions (or even just an obsolete name in this case) need not be tested in new tests, unless there is some edge case that affects the obsolete function or something (which is not here, this this just an alias)

File: test/functional/api/buffer_spec.lua [hunk 7]
```diff
@@ -35,8 +35,42 @@ describe('api/buf', function()
       -- There's always at least one line
       eq(1, curbuf_depr('line_count'))
     end)
-  end)
 
+    it('line_count has defined behaviour for unloaded buffers', function()
+      -- we'll need to know our bufnr for when it gets unloaded
+      local bufnr = request('nvim_buf_get_number', 0)
```
Comment (bfredl) [inline test/functional/api/buffer_spec.lua]: `nvim_buf_get_number` is deprecated, `nvim_get_current_buf` is preferred for this, or just use `curbuf()` without args.

**Comment by bfredl** [comment]
Good code changes, just some minor quibbles about docs and tests. :+1:

File: src/nvim/api/buffer.c (hunk 4 above)
Comment (phodge) [reply src/nvim/api/buffer.c]: I've updated the `nvim_buf_is_loaded()` docs to mention that you can also use `nvim_buf_line_count()`. I have also moved the comments on unloaded buffers to a new intro paragraph at the top of the `api-buffer` section.

File: test/functional/api/buffer_spec.lua [hunk 8]
```diff
@@ -35,8 +35,38 @@ describe('api/buf', function()
...
+    end)
+
+    it('get_lines has defined behaviour for unloaded buffers', function()
+      -- we'll need to know our bufnr for when it gets unloaded
+      local bufnr = buffer('get_number', 0)
+      -- replace the buffer contents with these three lines
+      buffer('set_lines', bufnr, 0, -1, 1, {"line1", "line2", "line3", "line4"})
+      -- confirm that getting lines works
+      eq({"line2", "line3"}, buffer('get_lines', bufnr, 1, 3, 1))
+      eq({"line2", "line3"}, request('buffer_get_lines', bufnr, 1, 3, 1))
```
Comment (bfredl) [inline test/functional/api/buffer_spec.lua]: this line can be removed

File: test/functional/api/buffer_spec.lua [hunk 9]
```diff
@@ -35,8 +35,38 @@ describe('api/buf', function()
       -- There's always at least one line
       eq(1, curbuf_depr('line_count'))
     end)
-  end)
 
+    it('line_count has defined behaviour for unloaded buffers', function()
+      -- we'll need to know our bufnr for when it gets unloaded
+      local bufnr = buffer('get_number', 0)
```
Comment (bfredl) [inline test/functional/api/buffer_spec.lua]: use `nvim('get_current_buf')` or `curbuf()` (one more case below)

File: scripts/gen_api_vimdoc.py [hunk 10]
```diff
@@ -413,10 +413,26 @@ def gen_docs(config):
...
 
     sections = {}
+    intros = {}
     sep = '=' * text_width
 
     base = os.path.join(out_dir, 'xml')
     dom = minidom.parse(os.path.join(base, 'index.xml'))
+
+    # generate docs for section intros
+    for compound in dom.getElementsByTagName('compound'):
```
Comment (bfredl) [inline scripts/gen_api_vimdoc.py]: nice improvement!

File: test/functional/api/buffer_spec.lua [hunk 11]
```diff
@@ -38,7 +38,7 @@ describe('api/buf', function()
 
     it('line_count has defined behaviour for unloaded buffers', function()
       -- we'll need to know our bufnr for when it gets unloaded
-      local bufnr = buffer('get_number', 0)
+      local bufnr = curbuf('get_number')
```
Comment (bfredl) [inline test/functional/api/buffer_spec.lua]: I meant just `curbuf()` with no args.

Comment (phodge) [reply test/functional/api/buffer_spec.lua]: I need the buffer number to use in `command('bunload! '..bufnr)`, but I can't figure out how to get the buffer number from the handle returned by `curbuf()` with no args.

Comment (bfredl) [reply test/functional/api/buffer_spec.lua]: The handle returned by `curbuf()` is the buffer number.

**Comment by bfredl** [comment]
LGTM. Squash and rebase?

**Comment by phodge** [comment]
@bfredl: I've rebased into a smaller set of commits. I'm not 100% certain what's going on with the build there - but I'm pretty sure the errors are unrelated. I think it's OK to merge now, but I don't have access to do that myself.

**Comment by bfredl** [comment]
Failures look unrelated, I restarted the failed builds.
